### PR TITLE
Better queue

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

File: src/server/pkg/pool/pool.go [hunk 1]
```diff
@@ -99,27 +104,32 @@ func (p *Pool) Do(ctx context.Context, f func(cc *grpc.ClientConn) error) error
...
+					if err != nil {
+						return err
+					}
+					mapConn.cc = cc
 					conn = mapConn
+					// We break because this conn has a count of 0 which we know
+					// we're not beating
+					break
+				} else {
+					mapCountCount := atomic.LoadInt64(&mapConn.count)
```
Comment (derekchiang) [inline src/server/pkg/pool/pool.go]: mapConnCount?

File: src/server/pkg/pool/pool.go [hunk 2]
```diff
@@ -99,27 +104,32 @@ func (p *Pool) Do(ctx context.Context, f func(cc *grpc.ClientConn) error) error
...
+					if mapCountCount < p.queueSize && (conn == nil || mapCountCount < atomic.LoadInt64(&conn.count)) {
+						conn = mapConn
+					}
 				}
 			}
+			if conn == nil {
+				p.connsCond.Wait()
```
Comment (derekchiang) [inline src/server/pkg/pool/pool.go]: I could be missing something, but seems like the only way for this to unblock is for new workers to be discovered.  But shouldn't we unblock also when a worker's queue is no longer full?

Comment (jdoliner) [reply src/server/pkg/pool/pool.go]: Ah, that's a very good point.

Comment (jdoliner) [reply src/server/pkg/pool/pool.go]: Should be fixed, by line 137 below.

**Pull request merged**